tools/make_patch.c

```diff
@@ -0,0 +1,475 @@
+	size_t namelen = strlen(name);
+	for (const struct Symbol *symbol = symbols; symbol; symbol = symbol->next) {
+		size_t symbolnamelen = strlen(symbol->name);
+		if (namelen > symbolnamelen) {
+			continue;
+		}
+		const char *symbolname = symbol->name;
+		if (name[0] == '.') {
+			// If `name` is a local label, compare it to the local part of `symbol->name`
+			symbolname += symbolnamelen - namelen;
+		}
+		if (!strcmp(symbolname, name)) {
+			return symbol;
+		}
+	}
+	error_exit("Error: Unknown symbol: %s\n", name);
+	return NULL;
+}
+
+const struct Symbol *find_symbol_cat(const struct Symbol *symbols, const char *prefix, const char *suffix) {
+	char *symname = xmalloc(strlen(prefix) + strlen(suffix) + 1);
+	strcpy(symname, prefix);
+	strcat(symname, suffix);
+	const struct Symbol *symbol = find_symbol(symbols, symname);
+	free(symname);
+	return symbol;
+}
+
+void parse_symbol_value(char *input, int *bank, int *address) {
+	char *buffer = xmalloc(strlen(input) + 1);
+	strcpy(buffer, input);
+
+	char *buffer2 = strchr(buffer, ':');
+	if (buffer2) {
+		// Parse symbol's bank:address
+		*buffer2++ = '\0';
+
+		char *endptr_bank, *endptr_address;
+		*bank = strtol(buffer, &endptr_bank, 16);
+		if (*bank > 0xff || *bank < 0) {
+			*bank = -1;
+		}
+		*address = strtol(buffer2, &endptr_address, 16);
+		if (*address > 0xffff || *address < 0) {
+			*address = -1;
+		}
+
+		if (endptr_bank != buffer + strlen(buffer) || endptr_address != buffer2 + strlen(buffer2)) {
+			error_exit("Error: Cannot parse bank+address: %s:%s\n", buffer, buffer2);
+		}
+	} else {
+		// Parse constant's value
+		*bank = 0;
+
+		char *endptr;
+		*address = strtol(buffer, &endptr, 16);
+		if (*address < 0) {
+			*address += 0x100;
+		}
+
+		if (endptr != buffer + strlen(buffer)) {
+			error_exit("Error: Cannot parse value: %s\n", buffer);
+		}
+	}
+
+	free(buffer);
+}
+
+struct Symbol *parse_symbols(const char *filename, struct Symbol **symbols) {
+	FILE *file = xfopen(filename, 'r');
+	struct Buffer *buffer = create_buffer(1);
+
+	int bank = 0;
+	int address = 0;
+	bool skip_line = false;
+
+	for (int c = getc(file); c != EOF; c = getc(file)) {
+		switch (c) {
+		case ';':
+		case '\r':
+		case '\n':
+			// This is the end of the symbol name
+			append_to_buffer(buffer, "");
+			append_symbol(symbols, buffer->data, bank, address);
+			// Clear the buffer and skip to the next line for another symbol
+			buffer->size = 0;
+			skip_line = true;
+			break;
+
+		case ' ':
+			// This is the end of the symbol value
+			append_to_buffer(buffer, "");
+			parse_symbol_value(buffer->data, &bank, &address);
+			// Clear the buffer for the symbol name
+			buffer->size = 0;
+			break;
+
+		default:
+			append_to_buffer(buffer, &c);
+		}
+
+		if (skip_line) {
+			// Skip to the next line
+			while (c != '\n' && c != '\r' && c != EOF) {
+				c = getc(file);
+			}
+			skip_line = false;
+		}
+	}
+
+	fclose(file);
+	free_buffer(buffer);
+	return *symbols;
```

Comment:
Why does this function return a pointer that it's modifying anyway?

Reply:
Same reason memcpy does.

Reply:
Actually because it used to be two functions, one which parsed the sym file and returned a newly allocated pointer, one which took the allocated pointer, parsed the constants file, and returned void. It doesn't need to return here.